### PR TITLE
chore(ci)!: Update tested Kubernetes versions

### PR DESCRIPTION
**Pull request by spencergilbert**
Signed-off-by: Spencer Gilbert <spencer.gilbert@gmail.com>

While Vector still _runs_ on kubernetes 1.14.x, as far as I can tell it's no longer offered/supported
by any major cloud provider. Right now #6564 is failing CI as we try to use a kubectl command that isn't
supported in 1.14.x (`rollout restart`).

This PR does the following:

- Drops 1.14.10

I'd also suggest consider adjusting our documentation around required/supported kubernetes versions. We'll also need to update and include more versions in a followup

### CODE REVIEW

File: .github/workflows/k8s_e2e.yml [hunk 1]
```diff
@@ -95,15 +95,16 @@ jobs:
...
             const minikube_version = [
-              "v1.15.1",
+              "v1.20.0",
             ]
             const kubernetes_version = [
-              { version: "v1.19.2", is_essential: true },
-              { version: "v1.18.9" },
-              { version: "v1.17.12" },
-              { version: "v1.16.15" }, // v1.16.13 is broken, see https://github.com/kubernetes/kubernetes/issues/93194
+              { version: "v1.21.2", is_essential: true },
```
Comment (jszwedko) [inline .github/workflows/k8s_e2e.yml:101]: Should we mark `is_essential: true` on all supported k8s versions?

**Review by jszwedko — COMMENTED** [review]
Should we update the docs with this change too? 👍 on supporting only the lowest available version on major cloud providers.

File: .github/workflows/k8s_e2e.yml (hunk 1 above)
Comment (spencergilbert) [reply .github/workflows/k8s_e2e.yml:101]: Maybe? It looks like we were doing it on the "current" version, marking them as essential would make less of a difference between the couple of labels we use to trigger e2e tests though

**Comment by spencergilbert** [comment]
> Should we update the docs with this change too? 👍 on supporting only the lowest available version on major cloud providers.

I wasn't sure if we wanted to - the chart still works on 1.14, we're no longer testing - I think it's a good idea to tighten the scope of supported versions, but that's also a larger discussion and (possibly) impactful to users/customers

**Comment by StephenWakely** [comment]
I'd say we need to update the docs. If we aren't testing it, we aren't supporting it.

**Comment by spencergilbert** [comment]
> I'd say we need to update the docs. If we aren't testing it, we aren't supporting it.

I'd defer to the group/leadership, but there's definitely some semantics here - Vector is still compatible with 1.14, we're not testing it actively, but we're also "best effort" not breaking existing compatibilities. Perhaps we could adopt language similar to minikube's:

> minikube follows the Kubernetes Version and Version Skew Support Policy, so we guarantee support for the latest build for the last 3 minor Kubernetes releases. When practical, minikube aims to support older releases as well so that users can emulate legacy environments.

File: .github/workflows/k8s_e2e.yml [hunk 2]
```diff
@@ -95,15 +95,16 @@ jobs:
...
+              { version: "v1.20.8" },
+              { version: "v1.19.12" },
+              { version: "v1.18.19" },
+              { version: "v1.17.17" },
+              { version: "v1.16.15" },
               { version: "v1.15.12" },
-              { version: "v1.14.10" },
```
Comment (binarylogic) [inline .github/workflows/k8s_e2e.yml:106]: We're only dropping support for `1.14.X`, right?

Comment (spencergilbert) [reply .github/workflows/k8s_e2e.yml:106]: Correct, this is removing the `1.14.X` test path - adding the more recent versions, and updating the patch versions

File: .github/workflows/k8s_e2e.yml (hunk 1 above)
Comment (jszwedko) [reply .github/workflows/k8s_e2e.yml:101]: It seems like this just controls the naming of the workflow job? We could drop it then as I don't see why we need it. I could see it if it was controlling which tests were "allowed failures" but currently there don't appear to be any.

**Comment by jszwedko** [comment]
> I'd say we need to update the docs. If we aren't testing it, we aren't supporting it.

I'm inclined to agree with this. If we aren't testing it, we aren't really supporting it.

My preference would be to support the lowest common denominator version for the major cloud providers.

**Comment by spencergilbert** [comment]
> > I'd say we need to update the docs. If we aren't testing it, we aren't supporting it.
> 
> I'm inclined to agree with this. If we aren't testing it, we aren't really supporting it.

👍 

> My preference would be to support the lowest common denominator version for the major cloud providers.

I think can figure this out in a following issue/rfc

> I do think we should update the docs with this PR.

I'll just update the docs to be `>=1.15` in this PR... though, do we need to consider this a breaking change?

File: .github/workflows/k8s_e2e.yml (hunk 1 above)
Comment (spencergilbert) [reply .github/workflows/k8s_e2e.yml:101]: I suspect when we've determined out "support period" (most recent 3/4, whatever versions) we should have them marked as essential 👍

**Comment by jszwedko** [comment]
> I'll just update the docs to be `>=1.15` in this PR... though, do we need to consider this a breaking change?

Yeah, this is a good question. I guess technically it should be a breaking change.

**Comment by spencergilbert** [comment]
I'll update it to be a breaking change, and I'll fix these patch versions - I went a little _too new_ for these patches, I think the failing ones are to be released next week 😅

**Comment by spencergilbert** [comment]
Tested the two failed jobs locally and they passed, so I'm guessing just flakey? Or maybe related to the runner env

**Comment by StephenWakely** [comment]
> Tested the two failed jobs locally and they passed, so I'm guessing just flakey?

I have a hunch about that flakey test. It's not to do with this PR. By creating the affinity pod in the same namespace as the test pod, sometimes the test picks up the logs from that pod instead. I'll get a fix.

**Comment by spencergilbert** [comment]
> > Tested the two failed jobs locally and they passed, so I'm guessing just flakey?
> 
> I have a hunch about that flakey test. It's not to do with this PR. By creating the affinity pod in the same namespace as the test pod, sometimes the test picks up the logs from that pod instead. I'll get a fix.

Talked a bit over slack we have _two_ sources of flakiness here, cri-o related and one in our tests @StephenWakely will fix

**Review by jszwedko — APPROVED** [review]

**Pull request merged**